### DISCUSSION TITLE
Comparison between kubekey and kubespray?

### DISCUSSION THREAD

**Discussion by yeya24** [kubesphere/kubekey#1178] [Q&A]
Hello community, we are using kubespray to provision/upgrade our k8s clusters in the bare metal environment. However, due to inflexibility reason and the cumbersome kubespray version dependency. We want to migrate to some operator based tool like kubekey.

Can I ask is there any comparison between kubekey and kubespray? Like features, supported versions, configurations, etc. If there is such comparison doc then it would be much easier for us to migrate. Thanks!

**Comment by pixiake** [comment]
KubeKey is the same as kuberspary, uses kubeadm to deploy clusters. And kubekey is written in go, so it doesn't need to rely on some basic software, such as python ansible. This also makes kubekey's installation speed faster, which is helpful for cluster installation in offline environment. 

Kubekey supports the use of [addons](https://github.com/kubesphere/kubekey/blob/master/docs/addons.md) to customize plugins when installing clusters.

Kubekey also supports convenient [offline cluster installation](https://github.com/kubesphere/kubekey/blob/master/docs/manifest_and_artifact.md).

In addition, kubekey also supports the operator mode. At present, only the cluster is abstracted. In the future, we plan to further optimize this mode, so as to realize the node abstraction.
